daml/ContingentClaims/Lifecycle.daml

```diff
@@ -3,83 +3,120 @@
+distM : Monad m
+      => Either (C a) (WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (C a)))
+      -> M m a
+distM (Left c) = pure (Left <$> (project c))
+distM (Right m) = fmap Right <$> m
+
+-- | Evaluate observables and skip branches for which predicates don't hold.
+-- This is useless onl it's own. Composed with other functions it adds laziness.
+lazily' : (Eq a, Monad m)
   => (Text -> Date -> m Decimal)
-  -> (C a -> C a -> C a)
-  -> C a -> Seed m a
+  -> C a -> Kleisli m Date (F a (Either (C a) (C a)))
+lazily' spot (When obs c) = do
+  predicate <- eval spot obs
+  pure if predicate then WhenF obs (Right c) else WhenF obs (Left c)
+lazily' spot (Cond obs c c') = do
+  predicate <- eval spot obs
+  if predicate then lazily' spot c else lazily' spot c'
+lazily' _ other = pure $ Right <$> project other
 
-lifecycle' _ _ Zero = pure ZeroF
-lifecycle' _ _ (One id) = do
+-- | Collects all `One` leaves replacing them with `Zero`s (to avoid double-spend).
+-- Calculates the product of any `Scale` factors between each `One` and the root node, logging these to the writer.
+-- This function **ignores** conditions; it traverses the entire tree.
+settle' : CanAbort m => (Text -> Date -> m Decimal) -> C a -> WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (C a))
+settle' _ (One asset) = do
   qty <- get
-  tell [(qty, id)]
-  pure ZeroF -- replace lifecycled node with zero, to avoid double-spend.
-lifecycle' _ _ (Give c) = do
-  modify negate 
-  pure $ GiveF (Right c)
-lifecycle' spot _ (When obs c) = do
-  predicate <- lift . lift $ eval spot obs
-  pure $ if predicate then WhenF obs (Right c) else WhenF obs (Left c)
-lifecycle' spot _ (Scale obs c) = do
+  tell [(qty, asset)]
+  pure ZeroF
+settle' _ (Give c) = do
+  modify negate
+  pure $ GiveF c
+settle' spot (Scale obs c) = do
   k <- lift . lift $ eval spot obs
-  modify (* k) 
-  pure $ ScaleF obs (Right c)
-lifecycle' spot choose (Or c c') = do
-  let chosen = choose c c'
-  if | chosen == c -> lifecycle' spot choose c --TODO: should we abort with nested choices?
-     | chosen == c' -> lifecycle' spot choose c'
-     | otherwise -> abort "lifecycle: Invalid branch election"
-lifecycle' _ _ (And c c') = pure $ AndF (Right c) (Right c')
-lifecycle' spot choose (Cond obs c c') = do
-  predicate <- lift . lift $ eval spot obs
-  if predicate then lifecycle' spot choose c else lifecycle' spot choose c'
+  modify (* k)
+  pure $ ScaleF obs c
+settle' _ (Or c c') = abort "settle: unexercised `Or`. Exercise it and then retry. "
+settle' _ other = pure $ project other
+
+-- | Used to make an election on `Or` or exercise an `Anytime`.
+--  `import` this `qualified`, to avoid clashes with `Prelude.exercise`.
+exercise : (Text -> Date -> m Decimal) -> C a -> C a -> Date -> C a
+exercise = undefined
+
+-- | Coalgebra for `exercise`.
+exercise' : Eq a => C a -> C a -> F a (C a)
+exercise' chosen (Or c _) | chosen == c = project c
```

Comment:
I think that one should not be allowed to exercise on `Or` | `Anytime` unless it is a root node. You may not own them.

Reply:
Here's a counter-example. How would you lifecycle this scenario:
```haskell
Scale 100.0 (c Or c')
```

Reply:
> You may not own them.

I've not got as far as to handle `Give` nodes, but the idea is that that is handled by a separate function, that you will compose this with. I.e. this function will only ever receive branches where it is the 'owner'.

Reply:
```
Scale 100.0 (c Or c') = (Scale 100.0 c) Or (Scale 100.0 c')
```

When we lifecycle a `Scale` we should move it to a terminal (next to a payment) position.

Reply:
> but the idea is that that is handled by a separate function..

I think this is where no one solution will handle all situations. For true trustless defi one should stop evaluation. For a custodian case, the custodian should mandate instructions from the client and a default action otherwise.

Performing the recursion explicitly allows us to control evaluation without the need for a special `Reader`

Reply:
See my previous comment. I believe we should be able to cover most situations now, as I have split the lifecycle and exercise into separate functions.
The `Lifecycle.exercise` now checks that only the bearer can make elections, accounting for `Give` nodes (even in nested expressions). The settlement function can be run by either party. These changes mean that we support any shape of `Claim`.